cmd/setup-etcd-environment/run.go

```diff
@@ -129,53 +164,82 @@ func runRunCmd(cmd *cobra.Command, args []string) error {
-	// etcd requires population of environment variable ETCD_UNSUPPORTED_ARCH at runtime.
-	// https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/supported-platform.md
-	arch := runtime.GOARCH
-	if strings.HasPrefix(arch, "s390") {
-		exportEnv["UNSUPPORTED_ARCH"] = arch
+	noExportEnv := map[string]string{
+		"IPV4_ADDRESS":      setupEnv.etcdIP,
+		"WILDCARD_DNS_NAME": fmt.Sprintf("*.%s", setupEnv.opts.discoverySRV),
 	}
-	if err := writeEnvironmentFile(exportEnv, out, true); err != nil {
-		return err
+	if setupEnv.etcdDNS != "" {
+		noExportEnv["DNS_NAME"] = setupEnv.etcdDNS
 	}
+	return writeEnvironmentFile(noExportEnv, out, false)
+}
 
-	return writeEnvironmentFile(map[string]string{
-		"IPV4_ADDRESS":      ip,
-		"DNS_NAME":          dns,
-		"WILDCARD_DNS_NAME": fmt.Sprintf("*.%s", runOpts.discoverySRV),
-	}, out, false)
+func (s *SetupEnv) getExportEnv() (map[string]string, error) {
+	if s.isEtcdScaling() {
+		return s.getOperatorEnv()
+	}
+	// initialize envs used to bootstrap etcd
+	return s.getBootstrapEnv()
 }
 
-func setExportEnv(etcdName, dns string) (map[string]string, error) {
-	exportEnv := make(map[string]string)
-	duration := 10 * time.Second
-	wait.PollInfinite(duration, func() (bool, error) {
-		if _, err := os.Stat("/var/run/secrets/kubernetes.io/serviceaccount/token"); os.IsNotExist(err) {
-			glog.Errorf("serviceaccount failed: %v", err)
-			return false, nil
-		}
-		return true, nil
-	})
+// getOperatorEnv returns a maping of ENV variables required to scale etcd via operator
+func (s *SetupEnv) getOperatorEnv() (map[string]string, error) {
+	env := make(map[string]string)
+	env["NAME"] = s.etcdName
+
+	memberList := s.getEtcdMemberList()
+	env["INITIAL_CLUSTER"] = strings.Join(memberList, ",")
+	env["INITIAL_CLUSTER_STATE"] = etcdInitialExisting
+
+	endpoints, err := s.getEtcdEndpoints()
+	if err != nil {
+		return nil, err
+	}
+	env["ENDPOINTS"] = strings.Join(endpoints, ",")
+	return env, nil
+}
 
-	clientConfig, err := rest.InClusterConfig()
+func (s *SetupEnv) getEtcdEndpoints() ([]string, error) {
+	ep, err := s.Client.CoreV1().Endpoints("openshift-etcd").Get("etcd", metav1.GetOptions{})
 	if err != nil {
-		panic(err.Error())
+		return nil, err
 	}
-	client, err := kubernetes.NewForConfig(clientConfig)
+	hostEtcdEndpoint, err := s.Client.CoreV1().Endpoints("openshift-etcd").Get("host-etcd", metav1.GetOptions{})
 	if err != nil {
-		return nil, fmt.Errorf("error creating client: %v", err)
+		return nil, err
+	}
+	if len(hostEtcdEndpoint.Subsets) != 1 {
+		return nil, fmt.Errorf("openshift-etcd/host-etcd endpoint subset length should be %d, found %d", 1, len(hostEtcdEndpoint.Subsets))
+	}
+
+	endpoints := make([]string, 0)
+	for _, member := range hostEtcdEndpoint.Subsets[0].Addresses {
+		// TODO validate IP?
+		if member.Hostname == etcdBootstrapName {
+			endpoints = append(endpoints, fmt.Sprintf("https://%s:2379", member.IP))
```

Comment:
will this fail if member.IP is empty/nil?  maybe check first

Reply:
yeah I think it is a loose assumption that has value, good catch.